### PR TITLE
Reduce database ops

### PR DESCRIPTION
**Pull request by epapbak**
# Description

- Retrieve previously notified reports right after getting the cluster list for new reports
- Filter previously notified reports to get only the ones that shouldn't be notified again (cooldown not finished yet)
- Update corresponding UTs

### Detailed description

- 01161d2 - Refactor `ReadLastNNotifiedRecords` to `ReadLastNotifiedRecordForClusterList`: 
  - With this change, we get all the previously notified records for the whole cluster list when the notification service is started, instead of one by one.
  - In a randomly selected production run, despite the improvement of only calling it for high impact rules, `shouldNotify` was called 10993 times. That means `ReadLastNNotifiedRecords`, which makes a call to the DB, was called 10993 times.

- 308b1ad - Reduce the number of records retrieved when getting the previously notified records for the whole cluster list
  - 01161d2 introduced a change that might cache too much data. With this additional change, the cooldown time is applied as a filter when retrieving the previously notified records, reducing greatly the number of records retrieved and making the cache way more efficient (depending on the configured cooldown time, of course).
  - The filter applied is `notified_at < NOW() - $cooldown`. So the records retrieved are the ones that are still on cooldown. If a new report contains an issue that is not in the retrieved records, it will be notified. If not, [we only compare the issues](https://github.com/RedHatInsights/ccx-notification-service/pull/194/commits/308b1adb15140a1e0c972c49405de65aac541b60#diff-e33b123d11a0b0d6a99929f7a43196756e6175a61f301aaf4cc14313e55ca072L83) and notify the ones not previously notified (as before).

Fixes [CCXDEV-8568](https://issues.redhat.com/browse/CCXDEV-8568)

## Type of change

- Breaking change (fix or feature that would cause existing functionality to not work as expected)
- This change requires a documentation update
- Refactor (refactoring code, removing useless files)

## Testing steps

- Manual testing
- Updated UTs
- Ran and verified BDD tests

## Checklist
* [ ] `make before_commit` passes
* [ ] updated documentation wherever necessary
* [x] added or modified tests if necessary
* [ ] updated schemas and validators in [insights-data-schemas](https://github.com/RedHatInsights/insights-data-schemas) in case of input/output change

### CODE REVIEW

File: config-devel.toml [hunk 1]
```diff
@@ -25,8 +25,8 @@ content_endpoint = "/api/v1/content"
 insights_advisor_url = "https://console.redhat.com/openshift/insights/advisor/clusters/{cluster_id}"
 cluster_details_uri = "https://console.redhat.com/openshift/details/{cluster_id}#insights"
 rule_details_uri = "https://console.redhat.com/openshift/details/{cluster_id}/insights/{module}/{error_key}"
-# Valid time units are "ns", "us" (or "µs"), "ms", "s", "m", "h".
-cooldown = "1m"
+# valid units are SQL epoch time units: months days hours minutes seconds"
+cooldown = "1 minutes"
```
Comment (tisnik) [inline config-devel.toml:29]: is the plural needed there?

File: differ/differ_cli_args_test.go [hunk 2]
```diff
@@ -0,0 +1,93 @@
+/*
+Copyright © 2021 Red Hat, Inc.
```
Comment (tisnik) [inline differ/differ_cli_args_test.go:2]: 2022

File: differ/export_test.go [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+// that need to be called from unit tests. Aliases should start with uppercase
+// letter because unit tests belong to different package.
+//
+// Please look into the following blogpost:
+// https://medium.com/@robiplus/golang-trick-export-for-test-aa16cbd7b8cd
+// to see why this trick is needed.
+
+func PushMetricsTest(metricsConf conf.MetricsConfiguration) {
+	pushMetrics(metricsConf)
+}
```
Comment (tisnik) [inline differ/export_test.go:35]: should not this file contains just name aliases?

File: differ/storage.go [hunk 4]
```diff
@@ -523,22 +534,27 @@ func (storage DBStorage) WriteNotificationRecordForCluster(
...
+	var orgIDs = make([]string, len(clusterEntries))
+	var clusterIDs = make([]string, len(clusterEntries))
+
+	for idx, entry := range clusterEntries {
+		orgIDs[idx] = strconv.FormatUint(uint64(entry.OrgID), 10)
+		clusterIDs[idx] = string(entry.ClusterName)
+	}
+
+	whereClause := fmt.Sprintf(` WHERE org_id IN (%v) AND cluster IN (%v) AND state = 1 `,
+		inClauseFromSlice(orgIDs), inClauseFromSlice(clusterIDs))
```
Comment (tisnik) [inline differ/storage.go:549]: does it even work for strings?

File: differ/export_test.go (hunk 3 above)
Comment (epapbak) [reply differ/export_test.go:35]: `PushMetrics` already exists in differ. I'll refactor =)

File: config-devel.toml (hunk 1 above)
Comment (epapbak) [reply config-devel.toml:29]: not needed. The unit can be any of `millennium, century, decade, year, month, week, day, hour, minute, second, millisecond, microsecond`; or the abbreviation (y, m, d, etc.,), or the plural forms (months, days, etc.).

But I think plural form, and "restricting" it from months to seconds makes sense when defining the cooldown =)

File: differ/storage.go (hunk 4 above)
Comment (epapbak) [reply differ/storage.go:549]: WDYM for strings?

I'm thinking of simplifying the function to only expect []string, so the caller function would be responsible for passing the correct format. Something like

```
func inClauseFromStringSlice(slice []string) string {
    if len(slice) == 0 {
        return ""
    }
    return "'" + strings.Join(slice, `','`) + `'`
}
```

File: config-devel.toml (hunk 1 above)
Comment (tisnik) [reply config-devel.toml:29]: ok

**Review by juandspy — APPROVED** [review]

**Review by tisnik — APPROVED** [review]
ok!

**Pull request merged**